File: src/main/java/kitchenpos/products/application/CreateProductService.java
package kitchenpos.products.application;

import kitchenpos.common.domain.vo.Name;
import kitchenpos.common.domain.infra.PurgomalumValidator;
import kitchenpos.products.tobe.domain.entity.Product;
import kitchenpos.products.tobe.domain.repository.ProductRepository;
import kitchenpos.common.domain.vo.Price;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;

@Service
public class CreateProductService {
    private final PurgomalumValidator purgomalumValidator;

    private final ProductRepository productRepository;

    public CreateProductService(
        final PurgomalumValidator purgomalumValidator,
        final ProductRepository productRepository
    ) {
        this.purgomalumValidator = purgomalumValidator;
        this.productRepository = productRepository;
    }

    @Transactional
    public Product create(final BigDecimal price, final String name) {
        final Price productPrice = new Price(price);
        final Name productName = new Name(name, purgomalumValidator);

        Product product = new Product(productName, productPrice);

        return productRepository.save(product);
    }
}

File: src/main/java/kitchenpos/products/tobe/domain/CreateProductService.kt
package kitchenpos.products.tobe.domain

import kitchenpos.products.domain.Product
import org.springframework.stereotype.Component
import java.math.BigDecimal
import java.util.*

@Component
class CreateProductService(
    private val productNameValidator: ProductNameValidator,
) {
    fun createProduct(
        price: BigDecimal?,
        name: String?,
    ): Product {
        ProductPriceValidator.requireNormalPrice(price)
        productNameValidator.requireNormalName(name)

        return Product().apply {
            this.id = UUID.randomUUID()
            this.name = name
            this.price = price
        }
    }
}
